### ISSUE TITLE
Stop simulation from within Node

### ISSUE THREAD

**Issue by arvoelke** [nengo/nengo#737]
I've encountered a situation where I want to run a simulation indefinitely, and decide when it ends it from within a Node (calculating the duration is expensive because it requires reading a few gigs of data). Currently I can do this in one of two ways:

 - Loop `sim.step` myself, checking a flag inside the node at each timestep
 - Raise an exception, and catch it on `sim.run`

The first is nice, but doesn't scale well to multiple such Nodes and the logic needs to be repeated for all users of that Node. The second would be nicer for users, as long as they don't need to wrap `sim.run` themselves.

So, I'm proposing that, similar to how an iterator will raise a `StopIteration`, nodes can raise a `StopSimulation` that will be nicely caught by the simulator. Support would also be required for passing None (infinity) as the number of steps.

**Comment by tcstewar** [comment]
+1 

On a somewhat related note, the ```nengo_spinnaker``` backend supports running forever (passing in ```None``` or maybe ```np.Inf``` to ```run()``` and they'd also like to be able to call ```sim.stop()``` from a separate thread to stop it running.

**Comment by tbekolay** [comment]
Also +1, this would be super nice. I also think it wouldn't be too difficult to implement, *but* it would have to be implemented by all backends (and we have at least 5 right now).

Should we have this in 2.1? I tentatively say yes.

**Comment by tcstewar** [comment]
I agree that this would be super nice.  And it should be a pretty simple change to the various backends (I think).  So let's try to get this into 2.1.  :)

One question, though, is what the syntax should look like for doing a "run forever" command.  Which of these should it be:

```python
sim.run()
sim.run(None)
sim.run(np.inf)
```

I think I'd argue for ```None```, but I'm fairly swayable....

Should we also support ```sim.stop()``` as part of this change?

Also, while we are working with this part of the API, the ```nengo_spinnaker``` backend needs a ```sim.close()``` to put the hardware back into a sane state at the end of a simulation.  I wouldn't be surprised if other hardware backends will need that too.  At the moment, ```nengo_gui``` handles this by calling ```sim.close()``` if it exists, but it might be nice to have it on all backends.

**Comment by jgosmann** [comment]
It would be nice to be able to do
```python
with Simulator(model) as sim:
    sim.run(10.)
```
and have closed to be called automatically.

**Comment by tcstewar** [comment]
That's actually exactly what the ```nengo_spinnaker``` currently supports as well.  They've set it up so you can do it either way (```__exit__``` just calls ```close```), which I think is correct.

**Comment by jgosmann** [comment]
:+1:

**Comment by tbekolay** [comment]
### Run forever syntax

Options:

1. `sim.run()`
2. `sim.run(None)`
3. `sim.run(np.inf)`

I think I prefer 1... there's a worry that people will do this without realizing it's running forever, but perhaps we can mitigate this by changing the args; right now we have `run(t=1.0)` which makes it obvious that it's running for 1.0 whatevers (seconds, it turns out). But if it were something like `run(t_end=1.0)` or `run(stop_after=1.0)` then passing nothing, and having it default to `None` would (hopefully) be pretty clear.

### `sim.close`, context manager

Yeah... I don't really like the context manager syntax for this (we already use it a lot, this is maybe too much?) It also feels unnecessary to force all simulators to implement this if not all of them need to do cleanup stuff on exit. Does registering an [`atexit` function](https://docs.python.org/2/library/atexit.html) in `Simulator.__init__` work for `nengo_spinnaker`?

### `sim.stop()`

The issue with `sim.stop()` is that it kind of gives the wrong impression of what's going in internally -- at least right now. Right now, we're using a synchronous blocking paradigm in which `sim.run` runs all the way though, and even if you do stop it, it's not expecting to be stopped. Even with the `StopSimulation` change, it's still blocking, and so a call to `sim.stop()` would have no effect unless it was called *inside* the `sim.run` loop (which is weird). Raising an exception still a little weird, but at least there's some indication that this is not a normal thing (it's exceptional, if you will :rainbow:).

However, if we switch to an asynchronous non-blocking paradigm, then something like `sim.stop` makes a lot of sense. But that would require a bit of an overhaul of the current simulator interface... it is probably a good idea to do though; it would make things like interacting with the GUI far more natural.

**Comment by tcstewar** [comment]
> Run forever syntax

I also like ```run()``` as an option, as implemented by defaulting the argument to ```None```.  However, I do worry about people who mean to call ```sim.run(10.0)``` and forget to pass in the argument (I've certainly done that myself).  In that situation, the simulation will just sit there running forever with no indication anything's wrong.  So I'm tempted to be explicit about the situation where I want it to run forever, especially since that's so rare.  So I'd like to require the ```None```.  I think you're right that changing the argument name might help here...  I've used ```limit``` in the past for this sort of thing, but I could sorta see ```stop_after```.

> sim.close, context manager

Unfortunately ```atexit``` doesn't work for SpiNNaker, since it needs to be called after each Simulation runs.  So in ```nengo_gui```, for example, it has to get called whenever it tears down an old simulation and starts up a new one.

That said, I'd totally be onboard with just have ```close()``` be required for now, and if we find we want to always allow the context manager approach as well, then that's very easy to add.  That said, it's only four lines to add this capability to any Simulator, so maybe it's worth adding anyway.

> sim.stop()

I'd be against switching over to an asynchronous non-blocking paradigm without a lot more thought.  It's a pretty big change, and avoiding threads when we can sounds like a good plan.  That said I see what you mean by ```sim.stop()``` giving the wrong impression.  What about ```sim.interrupt()```?  That makes it seem a bit more exceptional....  :)

**Comment by tbekolay** [comment]
> I do worry about people who mean to call sim.run(10.0) and forget to pass in the argument

Yeah, that's a worry. We can definitely provide some indication that we're running forever though; we already have the progress bar, which can be replaced by something that indicates we're running forever.

> it needs to be called after each Simulation runs

This seems like more of a SpiNNaker-specific thing than something that we should enforce for all simulators... Is it correct to say that you cannot have two `nengo_spinnaker.Simulator` instances at the same time, then? If so, it seems like something that should be enforced when a new `Simulator` is created (kind of like a singleton pattern, in a sense). E.g, in `Simulator.__init__` you check and see if a simulator currently exists, and if so, you print something to the console and tear it down.

It's definitely not a backend requirement that you can have multiple simulators, but I wouldn't want to add things to the API such that it looks like this is not a possibility, because it's certainly possible in the reference simulator.

> What about sim.interrupt()?

That is more exceptional! I'm more wondering: what are the situations in which you'd call this function? If it's inside the simulation, then it's probably in a node, and you'd probably want to just `raise StopSimulation` instead. If it's outside the simulation, then you are probably the one calling `sim.run`, and so you can't call `sim.interrupt`. The only situation it could occur is if you spin up some external watchdog thread... which means that you're trying to use Nengo in an asynchronous manner already? If I'm understanding things correctly...

**Comment by tcstewar** [comment]
> we already have the progress bar, which can be replaced by something that indicates we're running forever.

But we only have the progress bar on ```nengo```.  If we go with using the progress bar to indicate this then we'd want that on every backend.

> This seems like more of a SpiNNaker-specific thing than something that we should enforce for all simulators...

I'm pretty sure we'll need something like it for ```nengo_brainstorm``` and ```nengo_mpi``` and ```nengo_bluegene```.  Anything where you don't want two simulators accessing the same hardware at the same time will need some way of indicating that it's done with the hardware.  We can't just assume that it's done with it at the end of a ```sim.run(time)``` call, since we may way to call ```run``` again to continue the simulation.

> Is it correct to say that you cannot have two nengo_spinnaker.Simulator instances at the same time, then?

Nope, as I might have multiple SpiNNaker (or whatever hardware) boards plugged into the same machine.  So we might be able to get away with a global list of existing SpiNNaker objects and check them to see if one of them is using the same IP address as the one I want to talk to, and forcibly kill that one, but that seems more complex than having a ```close()``` function.

> That is more exceptional! I'm more wondering: what are the situations in which you'd call this function? 

It would always be called from some separate thread.  The particular use case right now would be to interrupt the SpiNNaker backend in nengo_gui.  Right now the nengo_spinnaker Simulator gets called repeatedly with a ```sim.run(10)``` and it can only get interrupted every 10 seconds, so the SpiNNaker simulation can take a few seconds to realize it has been stopped.  That said, we don't need an ```sim.interrupt()``` call in nengo_gui if it handles the proposed ```nengo.StopSimulation``` exception, since we're already making a Node that could generate that exception.  But it does seem like it'd nice to be able to generate such an exception without explicitly creating such a Node.

**Comment by jgosmann** [comment]
> I don't really like the context manager syntax for this

Isn't this like the canonical context manager example? We have some resource that should be closed in any case? I doesn't seem to me that
```python
sim = Simulator(model)
try:
    sim.run(10.)
finally:
    sim.close()
```
is any better.

> It also feels unnecessary to force all simulators to implement this if not all of them need to do cleanup stuff on exit

To me it seems more important that I can swap simulators without touching anything more than the line creating the simulator (and providing further configuration if needed). Also it doesn't seem much work to implement stubs for simulators that do not need cleanup or providing a base class that gives this implementation.

**Comment by tbekolay** [comment]
> We have some resource that should be closed in any case?

This is why I'm arguing against implementing `sim.close` in general. Because in most cases, we don't want to close the simulator after running for a certain amount of time, we want it to maintain its state so that we can call `sim.run` again. It would be just as annoying to have the opposite happen, where you blindly use the context manager then can no longer run your simulator object because it's destroyed itself.

> it seems more important that I can swap simulators without touching anything more than the line creating the simulator

I agree, but at the same time everything we add to `Simulator` then needs to be implemented on all simulators. So far, with what's been discussed in this thread, I think we have simulator-specific concerns that should be dealt with in simulator-specific ways (e.g., keeping a mapping from IP addresses to simulator instances and shutting down the old one when we start a new one).

**Comment by tcstewar** [comment]
> This is why I'm arguing against implementing sim.close in general.

I'm still pretty convinced that many backends will need a ```sim.close()```, but we do only have one right now, and so I could see putting off that discussion for now until we have a few more.  So we should probably spin that off to a separate Issue.  (I'd hoped it would be an uncontroversial change related to this Issue, but I guess not).

**Comment by studywolf** [comment]
+1 on calling sim.run multiple times, this is how i do everything because
it's way easier to have the nengo code inside the rest of my framework

**Comment by jgosmann** [comment]
> Because in most cases, we don't want to close the simulator after running for a certain amount of time, we want it to maintain its state so that we can call sim.run again

No one forces you to use the context manager in that case. You are still free to handle the close yourself. But I don't see what we gain by disallowing a common Python idiom in the simple cases where it would be useful. Actually, I kind of assume any object with a close method can be used as a context manager.
